### DISCUSSION TITLE
How to access "simplified names" of functions via java plugin API?

### DISCUSSION THREAD

**Discussion by utkonos** [NationalSecurityAgency/ghidra#6124] [Q&A] [answered]
A function symbol has the name `basic_string<char,struct_std::char_traits<char>,class_std::allocator<char>_>`. However, in the UI I can see what's called in Binary Ninja a simplified name: `std::basic_string<>::basic_string<>`.

When writing a Java plugin, how can I print this same string that is shown in the UI?

<img width="599" alt="Screenshot 2024-01-17 at 00 49 25" src="https://github.com/NationalSecurityAgency/ghidra/assets/2775739/894a9f1d-4f73-4ef4-a9b1-9fd282a4dd18">

**Comment by dragonmacher** [comment, marked as answer]
```java

ToolOptions fieldOptions = tool.getOptions(GhidraOptions.CATEGORY_BROWSER_FIELDS);
TemplateSimplifier simplifier = new TemplateSimplifier(fieldOptions);
...
String name = symbol.getName(true)
String simpleName = simplifier.simplify(name);

```

**Comment by utkonos** [comment]
@dragonmacher Thanks! Almost there! Where is `TemplateSimplifier`?

```none
ExportSymbols.java:50: error: cannot find symbol
        TemplateSimplifier simplifier = new TemplateSimplifier(fieldOptions);
```

<img width="590" alt="image" src="https://github.com/NationalSecurityAgency/ghidra/assets/2775739/9ea8e6cf-bdfe-4d47-9c88-6c852f6bfa3e">

**Reply by dragonmacher** [reply]
`ghidra.app.util.template.TemplateSimplifier`

**Comment by utkonos** [comment]
@dragonmacher Is there documenation for that? I can't find it. It works, however, so if undocumented, no problem.

**Reply by dragonmacher** [reply]
Here is the source.  The documentation is pretty sparse:

https://github.com/NationalSecurityAgency/ghidra/blob/master/Ghidra/Features/Base/src/main/java/ghidra/app/util/template/TemplateSimplifier.java

**Reply by utkonos** [reply]
Also, would adding "Simplified Name" as a column to the Symbol Table be a decent feature request? It could be unchecked in the Select Columns interface, so it only shows up if someone wants to see it.

**Reply by dragonmacher** [reply]
In glancing at the source, it appears you can use it without the tool options, configuring it as you like.  But, if you use the tool options, then as you configure the tool via the UI, those changes will be used by the simplifier.   Note though, it appears as though the simplifier does not listen for options changes.  So, to stay in sync with the options, you can either recreate the simplifier every time you need it, or use an `OptionsChangeListener` in your plugin.

Edit:
In glancing at the source, it appears you can use it without the tool options, configuring it as you like.  But, if you use the tool options, then as you configure the tool via the UI, those changes will be used by the simplifier.   Note though, it appears as though the simplifier does not listen for options changes.  So, to stay in sync with the options, you can either recreate the simplifier every time you need it, or use an `OptionsChangeListener` in your plugin.

**Reply by utkonos** [reply]
Yes, I removed the `fieldOptions` and it still works fine. Thanks!

**Reply by dragonmacher** [reply]
> Also, would adding "Simplified Name" as a column to the Symbol Table be a decent feature request? It could be unchecked in the Select Columns interface, so it only shows up if someone wants to see it.

It should be easy enough to add this column.  I imagine it could be useful for some binaries.  Feel free to create a new ticket for this request.

**Reply by utkonos** [reply]
Created: https://github.com/NationalSecurityAgency/ghidra/issues/6125

Thanks!!!
